8349211: Add support for intrusive trees to the utilities red-black tree

caspernorrbin opened the pull request:
Hi everyone,

The recently integrated red-black tree can be made more flexible by adding support of intrusive trees. In an intrusive tree, the user has full control over node allocation and placement instead of having the tree manage it internally.

Two key changes enable this feature:
1. Nodes can now be created outside of the tree's internal allocation mechanism, enabling users to allocate and prepare nodes before inserting them into the tree.
2. Cursors have been added to simplify navigation and iteration over the tree. These cursors are when inserting and removing nodes in an intrusive tree, where the internal tree allocator is not used. Additionally, cursors enable iteration over the tree and provide a convenient way to access node values.


Many of the auxiliary tree functions have been updated to use these new features, resulting in simplified and cleaned-up code. More tests have also been added to cover both new and existing functionality.

An example of how you could use the intrusive tree is found below:

```c++
    struct MyIntrusiveStructure {
      Node node; // The tree node is part of an external structure
      int data;

      MyIntrusiveStructure(int data, Node node) : node(node), data(data) {}
      Node* get_node() { return &node; }
      static MyIntrusiveStructure* cast_to_self(Node* node) { return (MyIntrusiveStructure*)node; }
    };

    Tree my_intrusive_tree;

    Cursor insert_cursor = my_intrusive_tree.cursor_find(0);
    Node insert_node = Node(0);

    // Custom allocation here is just malloc
    MyIntrusiveStructure* place = (MyIntrusiveStructure*)os::malloc(sizeof(MyIntrusiveStructure), mtTest);
    new (place) MyIntrusiveStructure(0, insert_node);

    my_intrusive_tree.insert_at_cursor(place->get_node(), insert_cursor);

    Cursor find_cursor = my_intrusive_tree.cursor_find(0);
    int found_data = MyIntrusiveStructure::cast_to_self(find_cursor.node())->data;
```


Please let me know any feedback or concerns!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8349211](https://bugs.openjdk.org/browse/JDK-8349211): Add support for intrusive trees to the utilities red-black tree (**Enhancement** - P4)


### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) Review applies to [7bd2b66b](https://git.openjdk.org/jdk/pull/23416/files/7bd2b66b16163353c63821e2bd41de58bf3ebc79)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23416/head:pull/23416` \
`$ git checkout pull/23416`

Update a local copy of the PR: \
`$ git checkout pull/23416` \
`$ git pull https://git.openjdk.org/jdk.git pull/23416/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23416`

View PR using the GUI difftool: \
`$ git pr show -t 23416`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23416.diff">https://git.openjdk.org/jdk/pull/23416.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23416#issuecomment-2630676635)
</details>

jdksjolen commented on test/hotspot/gtest/utilities/test_rbtree.cpp:
```diff
@@ -560,7 +748,7 @@ TEST_VM_F(RBTreeTest, VerifyItThroughStressTest) {
     }
   }
   { // Make a very large tree and verify at the end
-    struct Nothing {};
+  struct Nothing {};
```
Now can use `Empty` instead.

jdksjolen reviewed (commented):
Thanks,

A few comments. Still looking at this.

tstuefe commented:
Ping @stooke

tstuefe commented:
@caspernorrbin Thank you for this. Will make the RB tree much more useful to me.

About the example:

```
    Cursor insert_cursor = my_intrusive_tree.cursor_find(0);
    Node insert_node = Node(0);

    // Custom allocation here is just malloc
    MyIntrusiveStructure* place = (MyIntrusiveStructure*)os::malloc(sizeof(MyIntrusiveStructure), mtTest);
    new (place) MyIntrusiveStructure(0, insert_node);
```

Why not skip the useless insert_node creation? Let insert_at_cursor just accept uninitialized Node* memory, since it will initialize all Node members anyway?

caspernorrbin commented:
> Why not skip the useless insert_node creation? Let insert_at_cursor just accept uninitialized Node* memory, since it will initialize all Node members anyway?
> 

As of now, the cursor doesn't track to what key its pointing to.  All `Node` members _but_ the key are initialized, so the key still needs to be specified. Thats what the `0` in `Node(0)` is. I could change this to have that also stored inside the cursor, so we can avoid creating the node first.

> could you massage this patch a bit to reduce the delta to the last version? That is a good idea in general (I usually do a manual minimize-delta sweep before I undraft a PR for review). A lot seems to be code movement at first glance.
> 

A lot is new or rewritten so git diff had a hard time picking up related changes. I tried my best to order and group the changes so functionality overlaps.

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -66,7 +68,22 @@ class RBTree {
     const K& key() const { return _key; }
     V& val() { return _value; }
-    const V& val() const { return _value; }
+    V& val() const { return _value; }
```
Oops, fixed now :)

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -30,6 +30,9 @@
 #include "utilities/globalDefinitions.hpp"
 #include <type_traits>
 
+struct Empty {};
```
`Empty` was used as the value type for the intrusive tree, but I discovered that it didn't quite work as expected, because `sizeof(Empty) == 1` due to it requiring a unique address. This means that we would waste space in a lot of cases. For example, 8-byte keys would lead to 40-byte `RBNode`s despite storing only three 8-byte pointers and one 8-byte key.

I explored an alternative approach by adding the option to use void as the value type instead, and removing the need for a value member altogether. By using a base class containing only the value and using conditional inheritance from either it or `Empty` (which benefits from [empty base optimization](https://en.cppreference.com/w/cpp/language/ebo)), we can have a zero-size overhead.

Code-wise, this solution doesn't look as clean. We need added templating and `ENABLE_IF`s for functions with value references (since `void&` doesn't work). Another positive however is that this enables key-only red-black trees for scenarios where values aren't necessary.

Let me know what you think :)

jdksjolen replied:
Yeah, it's a bit "ugly", but I'm willing to pay the price for a little bit of complexity in order to have this feature work as expected. Please add "empty base optimization" as a phrase in a comment regarding this, so that people can find out on their own why the code looks as it does.

caspernorrbin commented:
I changed the way `insert_at_cursor` so it initializes the node so we can have completely uninitialized memory. To do this, the key used when creating the cursor is now stored in it.

The example would now look something like this:
```c++
    struct MyIntrusiveStructure {
      Node node; // The tree node is part of an external structure
      int data;

      MyIntrusiveStructure(int data) : data(data) {}
      Node* get_node() { return &node; }
      static MyIntrusiveStructure* cast_to_self(Node* node) { return (MyIntrusiveStructure*)node; }
    };

    Tree my_intrusive_tree;
    Cursor insert_cursor = my_intrusive_tree.cursor_find(0);

    // Custom allocation here is just malloc
    MyIntrusiveStructure* place = (MyIntrusiveStructure*)os::malloc(sizeof(MyIntrusiveStructure), mtTest);
    new (place) MyIntrusiveStructure(123);

    my_intrusive_tree.insert_at_cursor(place->get_node(), insert_cursor);

    Cursor find_cursor = my_intrusive_tree.cursor_find(0);
    int found_data = MyIntrusiveStructure::cast_to_self(find_cursor.node())->data;
```

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -161,30 +218,111 @@ class RBTree {
+  // For all nodes with key > old_node, must also have key > new_node
+  void replace_at_cursor(RBNode* new_node, const Cursor& cursor);
+
+  // Finds the value of the node associated with the given key.
+  V* find(const K& key) {
+    Cursor cursor = cursor_find(key);
+    return cursor.found() ? &cursor.node()->_value : nullptr;
+  }
+
+  V* find(const K& key) const {
+    const Cursor cursor = cursor_find(key);
+    return cursor.found() ? &cursor.node()->_value : nullptr;
+  }
```
Are these important to have?

caspernorrbin replied:
Depends on what you mean by important. I think it's useful for those who want to interact with the tree on a higher level, by for example only using `upsert(k,v)`, `remove(k)`, and `find(k)`. Of course, I could remove this and force people to use cursors instead, but that feels unnecessary, especially if they don't need cursors otherwise.

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -161,30 +218,111 @@ class RBTree {
+    return cursor.node();
+  }
+
+  // Inserts a node with the given key into the tree,
+  // does nothing if the key already exist.
+  void upsert(const K& key) {
```
`upsert` is a bad name for function, as it is a portmanteau of "update" and "insert", indicating that this should change something if the key is found. What's the goal with this function? It should probably return the allocated node if it's to be useful.

caspernorrbin replied:
This follows the pattern of the other value-less functions keeping the name, but I agree that it's less logical here. Would one `insert(k)` and one `upsert(k, v)` be a good solution?

jdksjolen replied:
Yes!

jdksjolen commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -188,62 +188,49 @@ inline void RBTree<K, V, COMPARATOR, ALLOCATOR>::RBNode::verify(
-  RBNode* curr = _root;
-  while (curr != nullptr) {
-    const int key_cmp_k = COMPARATOR::cmp(key, curr->key());
+inline const typename RBTree<K, V, COMPARATOR, ALLOCATOR>::Cursor
+RBTree<K, V, COMPARATOR, ALLOCATOR>::cursor_find(const K& key) const {
```
```c++
using Tree = RBTree<K, V, COMPARATOR, ALLOCATOR:
using Cur = typename RBTree<K, V, COMPARATOR, ALLOCATOR>::Cursor;
```

Though I'm pretty sure Thomas gave you a `typedef` for `Tree`, I don't think the `Cur` can be done with a `typedef`.

caspernorrbin replied:
I don't think we can access `TreeType` here in any meaningful way, since that is a part of the tree class and we're outside the class here, and we would still need to specify the template parameters.  But please correct me if I'm wrong :)

jdksjolen replied:
I don't know :-), you might be right!

jdksjolen commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -188,62 +188,49 @@ inline void RBTree<K, V, COMPARATOR, ALLOCATOR>::RBNode::verify(
+    const int key_cmp_k = COMPARATOR::cmp(key, (*insert_location)->key());
 
     if (key_cmp_k == 0) {
-      return curr;
-    } else if (key_cmp_k < 0) {
-      curr = curr->_left;
+      break;
+    }
+
+    parent = *insert_location;
+    if (key_cmp_k < 0) {
+      insert_location = &((*insert_location)->_left);
     } else {
-      curr = curr->_right;
+      insert_location = &((*insert_location)->_right);
     }
```
```
while (*insert_location != nullptr) {
  auto find_a_Good_Name = *insert_location; // And replace the usage sites
}
```

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -116,21 +144,54 @@ class RBTree {
+  // Represents the location of a (would be) node in the tree.
+  // If a cursor is valid (valid() == true) it points somewhere in the tree.
+  // If the cursor points to an existing node (found() == true), node() can be used to access that node,
+  // Otherwise nullptr is returned, regardless if the node is valid or not.
```
Style: "nullptr" should be "null" here.

jdksjolen replied:
"Otherwise", but the above line has a comma.  What can be used to access that node? You don't say :-).

jdksjolen commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -116,21 +144,54 @@ class RBTree {
+    Cursor(RBNode* const* insert_location, RBNode* parent, const K& key)
+        : _insert_location((RBNode**)insert_location), _parent(parent), _key(key) {}
+
+  public:
+    bool valid() const { return _insert_location != nullptr; }
+    bool found() const { return *_insert_location != nullptr; }
+    RBNode* node() { return _insert_location == nullptr ? nullptr : *_insert_location; }
+    RBNode* node() const { return _insert_location == nullptr ? nullptr : *_insert_location; }
```
Is there any case where I don't need to check the validity of the cursor? That is, do I ever want to use `node()` without first calling `valid()` or afterwards checking whether the returned value was null?

If the answer to that is: "No, there is no such case", then we shouldn't return null on `!valid()` node. We should instead add `assert(valid(), "must be");". If the answer is yes, then could you please tell me what that situation is :P?

caspernorrbin replied:
It's been useful in a couple of places, where we want "the node or nullptr otherwise", since you get the valid check and the node in one. A few examples where we don't check validity:
In `visit_range_in_order`, we iterate until the node (or nullptr) is reached.
In `upsert`, we extract the node node into a local variable and either modify the node or reuse the variable.
In `closest_gt`, we return the next node from the cursor, which could be invalid.

jdksjolen replied:
OK, then leave it as is :-).

caspernorrbin commented on src/hotspot/share/utilities/rbTree.hpp:
```diff
@@ -161,30 +218,111 @@ class RBTree {
+  // There must not exist any node with the same key
+  // For all nodes with key < old_node, must also have key < new_node
+  // For all nodes with key > old_node, must also have key > new_node
+  void replace_at_cursor(RBNode* new_node, const Cursor& cursor);
```
We already call `verify_self` in the implementation :)

jdksjolen reviewed (approved):
Alright,

I'm okay with this. As all the internals now use cursors we basically get the intrusive-style API tested for free.

Thanks for your efforts on this!

tstuefe reviewed (commented):
Hi @caspernorrbin,

Sorry for the delay. I tried to use the tree in a tiny small allocator (basically what I plan to do in Metaspace and in other places), and though it is now possible in principle I think we can simplify things and make them more user-friendly. Here my findings:

- It would be good to have RBNode simplified and defined outside of the tree. Possibly even in a different header. I can see RBNode being used in places where I don't know exactly what tree it goes into. It could even go into multiple trees at the same time or at different times (so the data structure would have either multiple RBNode inlined or a single one that gets repurposed for different trees).

- In that line, it would be good to have the key in RBNode being mutable. Having it const means I am forced to write constructors for containing structures. That is cumbersome.

- To me, it seems the code could be a lot simpler if you were to just use standard subclassing (AbstractRBTree->(NonIntrusiveRBTree|IntrusiveRBTree) etc. All these `std::is_same` are a bit much. There would also be no need for the `RBTreeNoopAllocator`. The vtable tax you'd pay won't matter much in reality since I don't foresee many cases where the specific tree type is not known. 

- I found I had little need for cursors at all. Mostly, they just got in the way. Cursor exists to modify tree structure, but why would I ever do that manually? It is different with simple structures like linked lists, but here the tree balances itself, so it has the last say about its structure anyway. I would be perfectly happy with just the simple ability to add/remove nodes manually, use nodes to find nearby nodes (as in, nodes of nearby keys), iterate nodes with a functor etc. 

- The few cases I needed a cursor it was because the API forced me to (e.g. when removing a node from the tree). With insertion, it got very weird. So I have an RBNode*, want to insert it into the tree, now I need an empty Cursor to do that? So, I create an empty cursor with that key, then use that with insert_at_cursor? Why?

- Let's say I already have a nearby node (result of closest_gt, for instance), but it does not satisfy me, so I add a new one. For that I need to call normal insert, so the search is done all over again (see my remark above above). It would be good if we could have an insertion with a node as an insertion hint.

- I found that I miss a closest_ge (greater or equal). Please add that. It's really needed for memory management trees.

- In closest_gt() etc, please extend the comments to say what the behavior is when the node is not found. I assume null is returned?

I'll continue the work later (have vacation next week).

xmas92 commented:
Hi @tstuefe

I'll give some background before I discuss this implementation.

I think a lot of the design here has been taken from or is a consequence of our work in ZGC where we are rewriting parts of our internal memory management. The initial work used an early version of @caspernorrbin's non-intrusive RB-tree. We wanted to reduce the amount of work required under our allocation lock, so we had the idea of making the RB-tree intrusive and permit 'updates' where we can modify a node, and/or replace it as long as we do not break the normal strict node ordering invariant. So I built ZIntrusiveRBTree as an experiment and it worked out very well from a performance perspective. The success of us using an intrusive tree along with your wishes to also have an intrusive tree, made @jdksjolen ask us if we could use Casper's tree if it enabled support for a intrusive nodes. We would very much appreciate using a shared implementation.

So I stopped iterating on the `ZIntrusiveRBTree`. But here is the implementation:
https://github.com/xmas92/jdk/blob/757daf9bac98e7cb04121664d2e562f31e817faa/src/hotspot/share/gc/z/zIntrusiveRBTree.hpp
https://github.com/xmas92/jdk/blob/757daf9bac98e7cb04121664d2e562f31e817faa/src/hotspot/share/gc/z/zIntrusiveRBTree.inline.hpp

> It would be good to have RBNode simplified and defined outside of the tree. Possibly even in a different header. I can see RBNode being used in places where I don't know exactly what tree it goes into. It could even go into multiple trees at the same time or at different times (so the data structure would have either multiple RBNode inlined or a single one that gets repurposed for different trees).

I would prefer this as well, at least in the `ZIntrusiveRBTree` at some point I found it cumbersome having the node be an inner class of the tree, so I moved it out of the tree.

> In that line, it would be good to have the key in RBNode being mutable. Having it const means I am forced to write constructors for containing structures. That is cumbersome.

Our use case would need this, or rather in our use case the key would probably be a dummy. In 'ZIntrusiveRBTree' we materialize the `Key` from `Node*`. Essentially we are keeping tracked of memory regions, so the virtual key is a Memory Range. https://github.com/xmas92/jdk/blob/757daf9bac98e7cb04121664d2e562f31e817faa/src/hotspot/share/gc/z/zMappedCache.cpp#L42-L49

And our lookups uses the actual address (a smaller key) to find where to insert a value. So our compare is always Compare(Key, Node*) (the Compare(Node*, Node*) is only used for verification).
https://github.com/xmas92/jdk/blob/757daf9bac98e7cb04121664d2e562f31e817faa/src/hotspot/share/gc/z/zMappedCache.cpp#L95-L107

STL has a similar concept with `template< class K > iterator find( const K& x );` which allows for a different type to be used for the lookup. But the whole "the key is a function of the `Node*`" is a completely intrusive property, so it creates a discrepancy between the two trees types.

We can use this list as it is, as we can just use the address of the key, but it relies on that implementation hands out a reference to the actual key in the node and not a copy. It feels flaky (and wastes bytes). I think that in the intrusive case having the Comparator compare the key to a `Node*` is preferable. And just having the intrusive node just carry the tree structure data, and let the containing object worry about the key and the value. Or maybe there is some more elegant API I am missing. Could also use meta-programming to select the (Key, Node*) operator if it exist and use (Key, Key) otherwise. 

> I found I had little need for cursors at all. Mostly, they just got in the way. Cursor exists to modify tree structure, but why would I ever do that manually? It is different with simple structures like linked lists, but here the tree balances itself, so it has the last say about its structure anyway. I would be perfectly happy with just the simple ability to add/remove nodes manually, use nodes to find nearby nodes (as in, nodes of nearby keys), iterate nodes with a functor etc.
> The few cases I needed a cursor it was because the API forced me to (e.g. when removing a node from the tree). With insertion, it got very weird. So I have an RBNode*, want to insert it into the tree, now I need an empty Cursor to do that? So, I create an empty cursor with that key, then use that with insert_at_cursor? Why?
> Let's say I already have a nearby node (result of closest_gt, for instance), but it does not satisfy me, so I add a new one. For that I need to call normal insert, so the search is done all over again (see my remark above above). It would be good if we could have an insertion with a node as an insertion hint.

I was in the process of trying to use iterators instead of the cursors when I stopped working on the `ZIntrusiveRBTree`. It works right now for our use case. However there are currently may ways to do the same thing, you can use both Nodes, Cursors and Iterators to walk the tree. Both nodes and iterators track a location in the tree while the cursors have more contextual information, as they also track the insert location.

The cost of using insertion hints is that it requires at least on redundant compare, which should be irrelevant. And requires some extra handling of the Tree::end() / nullptr hint, the easiest way I can think of is to have the tree keeps track of the right most node so it can find the insert location fast. 

Iterator hints are safer to use externally, as they are just hints, the insertion will still respect the comparator function. While with the cursor, you can insert an invalid node somewhere.

As long as we do not care to about wether insert succeeds or not, I think the API where you consume an iterator hint and produce a new iterator as a result could end up useful. Similar to `std::set` / `std::map` API.

The main difference with `Node*`s/iterators over cursors is that one would use `gt` over `find` as `find`. Which I think is more natural in many places as it describes. 
```c++
  // Insert or merge
  auto it = tree.gt(key);
  if (should_merge(*it, node))
    tree.replace(it, merge_node(*it, node));
  else
    tree.insert(it, node);

  // Remove all nodes >=key
  for (auto it = tree.find(key); 
       it != tree.end();
       it = tree.remove(it));
```

The nice thing with iterators over `Node*`s is that they are more versatile, and we it allows us to remove traversing the tree through `Node*`'s public API, as such they are just some space for the tree to store its data. And I think we also can get a safer interface and type system. You use `Node*` when referring to something that may not be part of the tree and you use iterators when referring to something that is in the tree (sans iterator invalidation, or using the iterator on a different tree). I think I mentioned in #22360 that implementing lightweight iterators (effectively just a fancy `Node*`, in `ZIntrusiveRBTree` I also keep track of the tree to allow for stronger asserts and bidirectionally) allows for a lot of the APIs and the implementation to be implemented in terms of the iterators.

I think regardless of what is used we should have API overloads that do not require using a hint or cursor. `Tree::remove(Node*)` and `Tree::insert(Node*)` (or `Tree::insert(Key, Node*)`).

Just my 2 cents.

caspernorrbin commented:
I just pushed a change addressing some of the points mentioned:
- Separated `RBNode` into its own class
- Separated the 'normal' `RBTree` (the parts using an allocator) to its own subclass, with the the intrusive tree essentially becoming the base.
- Added `insert_node(Node)` and `remove(Node)` so that you can use the intrusive parts without necessary having to use cursors.
- Added an optional `hint_node` parameter to insert and lookup functions that if given, starts searching for the intended node/place starting at that node instead of the tree root

Any feedback would be appreciated!
Thank you both for the comprehensive comments, I'll continue working on this.

caspernorrbin commented:
Just pushed another change, this time separating the node class into `IntrusiveRBNode` and `RBNode<K, V>`. `IntrusiveRBNode` stores the tree pointers and is used for the intrusive tree, and no longer needs to be templated. `RBNode<K, V>` is used for the normal tree and also stores a key and a value. This means that the key is now stored outside the node for intrusive trees. To allow for this, I changed the compare function for intrusive trees to `cmp(K a, const IntrusiveRBNode* b)`. For the normal tree, both `cmp(K a, const RBNode<K, V>* b)` and `cmp(K a, K b)` can be used. With all the changes so far, the original example would now look something like this:

```c++
    struct MyIntrusiveStructure {
      IntrusiveRBNode node; // The tree node is part of an external structure
      int key;
      int data;

      MyIntrusiveStructure(int key, int data) : key(key), data(data) {}
      IntrusiveRBNode* get_node() { return &node; }
      static MyIntrusiveStructure* cast_to_self(IntrusiveRBNode* node) { return (MyIntrusiveStructure*)node; }
    };

    struct IntrusiveCmp {
      static int cmp(int a, const IntrusiveTreeNode* b) {
        return a - IntrusiveHolder::cast_to_self(b)->key;
      }
    };

    IntrusiveRBTree<int, IntrusiveCmp> my_intrusive_tree;
    const int key = 0;

    // Custom allocation here is just malloc
    MyIntrusiveStructure* place = (MyIntrusiveStructure*)os::malloc(sizeof(MyIntrusiveStructure), mtTest);
    new (place) MyIntrusiveStructure(key, 123);

    my_intrusive_tree.insert(key, place->get_node());

    IntrusiveRBNode* found_node = my_intrusive_tree.find_node(key);
    int found_data = MyIntrusiveStructure::cast_to_self(found_node)->data;
```

The key changes are:
- The key is stored in `MyIntrusiveStructure` and can be modified/changed whenever.
- No node constructor is needed and you simply need to pass the node pointer to `insert`.
- No cursors are needed (although you could still use them) to insert/lookup/delete nodes.

caspernorrbin commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -475,46 +481,143 @@ inline void RBTree<K, V, COMPARATOR, ALLOCATOR>::remove(RBNode* node) {
+  if (old_node == new_node) {
+    return;
+  }
+
```
The `cmp(const NodeType* a, const NodeType* b)` function is identical to the one you needed to give as a template to `verify_self` if you're using intrusive trees. I made that function be a part of the `COMPARATOR` template instead. Now, you supply a second `cmp` in `COMPARATOR` used for verification only, and that is used in both `verify_self` and `replace_at_cursor`.

xmas92 commented on src/hotspot/share/utilities/rbTree.inline.hpp:
```diff
@@ -605,7 +613,12 @@ inline void AbstractRBTree<K, NodeType, COMPARATOR>::visit_range_in_order(const
+  if (start != nullptr) {
+    assert_leq(from, start);
+    assert_geq(to, start);
+  }
```
Not sure if we should add an else branch here where we assert end == nullptr / end == start. But given that we will more than likely just crash when reading `start->next()`, it does not matter to much. Regardless of any assert, a bad interval will crash early.

caspernorrbin replied:
Added an assert just in case.

xmas92 reviewed (approved):
lgtm.

Thansk!

jdksjolen reviewed (approved):
>ACTION: main -- Error. Program `/home/runner/work/jdk/jdk/bundles/jdk/jdk-25/fastdebug/bin/java' timed out (timeout set to 480000ms, elapsed time including timeout handling was 520263ms).
REASON: User specified action: run main/othervm -Xmx1g -Xms1g -Xlog:gc -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions -XX:ShenandoahTargetNumRegions=2048 -XX:+UseShenandoahGC -XX:ShenandoahGCHeuristics=adaptive -XX:ShenandoahGCMode=generational -XX:+ShenandoahVerify TestAllocHumongousFragment 

I think it's safe to say that this crash is unrelated to  your changes.

Still LGTM.

caspernorrbin commented:
Thank you everyone for your time on this!

/integrate
